### PR TITLE
Constant Folding: Only apply size limit if size of tensor is increasing

### LINKED ISSUE #24083: Constant folding doesn't fold weights that are using weight normalization
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 1.12.0
- Python version: 3.5.2
- Bazel version (if compiling from source): 0.15.0?
- GCC/Compiler version (if compiling from source):  5.4.0
- CUDA/cuDNN version: CUDA 10.0.130, CUDNN 7.4.1.5
- GPU model and memory: TITAN V 10956 MB

**Describe the current behavior**
I have a 1D convolution layer whose weights are using weight normalization (Salimans & Kingma, 2016). The weights to the convolution are processed using the following equation: `w = g * v/2-norm(v)`. After running the constfold optimizer, these operations on the weights are still present, even though they should be folded.

**Describe the expected behavior**
I would expect the normalization ops to be folding into the weights, so that only one Const/read node remains.

**Code to reproduce the issue**
```python
import tensorflow as tf
import math
from tensorflow.core.protobuf import meta_graph_pb2
from tensorflow.core.protobuf import rewriter_config_pb2
from tensorflow.python.grappler import tf_optimizer

def conv_graph(x, output_name='output', kernel_width=3, in_dim=1024, out_dim=1024):
  """Applies convolution with gated linear units on x.
  https://github.com/NVIDIA/OpenSeq2Seq/blob/master/open_seq2seq/parts/convs2s/conv_wn_layer.py
    Args:
      x: A float32 tensor with shape [batch_size, length, in_dim]
    Returns:
      float32 tensor with shape [batch_size, length, out_dim].
  """
  # Define Variables
  conv_out_size = 2 * out_dim
  V_std = math.sqrt(4.0 * 0.8 / (kernel_width * in_dim))
  V = tf.get_variable('V', shape=[kernel_width, in_dim, conv_out_size],
                      initializer=tf.random_normal_initializer(mean=0, stddev=V_std),
                      trainable=True)
  V_norm = tf.norm(V.initialized_value(), axis=[0, 1])
  g = tf.get_variable('g', initializer=V_norm, trainable=True)
  W = tf.reshape(g, [1, 1, conv_out_size]) * tf.nn.l2_normalize(V, [0, 1])

  output = tf.nn.conv1d(value=x, filters=W, stride=1, padding="VALID")
  output = tf.identity(output, name=output_name)
  return output

def apply_constfold(frozen_graph, output_nodes):
  graph = tf.Graph()
  with graph.as_default():
    tf.import_graph_def(frozen_graph, name="")
  grappler_meta_graph_def = tf.train.export_meta_graph(graph_def=graph.as_graph_def(add_shapes=True), graph=graph)

  _to_bytes = lambda s: s.encode("utf-8", errors="surrogateescape")
  output_collection = meta_graph_pb2.CollectionDef()
  output_list = output_collection.node_list.value
  for i in output_nodes:
    if isinstance(i, tf.Tensor):
      output_list.append(_to_bytes(i.name))
    else:
      output_list.append(_to_bytes(i))
  # TODO(laigd): use another key as the outputs are really not train_op.
  grappler_meta_graph_def.collection_def["train_op"].CopyFrom(output_collection)
  rewriter_config = rewriter_config_pb2.RewriterConfig()
  rewriter_config.optimizers.extend(["constfold"])

  session_config_with_trt = tf.ConfigProto()
  session_config_with_trt.graph_options.rewrite_options.CopyFrom(
      rewriter_config)
  frozen_graph = tf_optimizer.OptimizeGraph(session_config_with_trt, grappler_meta_graph_def, graph_id=b"tf_graph")
  return frozen_graph

if __name__ == '__main__':
  with tf.Graph().as_default():
    # Create graph
    x = tf.placeholder(dtype=tf.float32, shape=(None, None, 1024), name='input')
    y = conv_graph(x)
    # Initialize
    with tf.Session() as sess:
      sess.run(tf.global_variables_initializer())
      # Freeze graph
      frozen_graph = tf.graph_util.convert_variables_to_constants(
          sess,
          sess.graph_def,
          output_node_names=['output'])

  print('Nodes before:')
  [print(n.name, n.op) for n in frozen_graph.node]

  # const folding
  frozen_graph = apply_constfold(frozen_graph, output_nodes=['output'])

  print('----------------------------------------')
  print('Nodes after:')
  [print(n.name, n.op) for n in frozen_graph.node]
```
**Output of script**
```
Nodes before:
input
V
V/read
g
g/read
Reshape/shape
Reshape
l2_normalize/Square
l2_normalize/Sum/reduction_indices
l2_normalize/Sum
l2_normalize/Maximum/y
l2_normalize/Maximum
l2_normalize/Rsqrt
l2_normalize
mul
conv1d/ExpandDims/dim
conv1d/ExpandDims
conv1d/ExpandDims_1/dim
conv1d/ExpandDims_1
conv1d/Conv2D
conv1d/Squeeze
output
----------------------------------------
Nodes after:
input
V
Reshape
l2_normalize/Sum/reduction_indices
l2_normalize/Maximum/y
conv1d/ExpandDims/dim
conv1d/ExpandDims_1/dim
V/read
conv1d/ExpandDims
l2_normalize/Square
l2_normalize/Sum
l2_normalize/Maximum
l2_normalize/Rsqrt
l2_normalize
mul
conv1d/ExpandDims_1
conv1d/Conv2D
conv1d/Squeeze
output
```

Edit: updated repro for 1.13

### PR DESCRIPTION
**Pull request by trevor-m**
Many kernels have sizes larger than the constfolding limit of 10mb. Many networks use identity ops or pointwise scaling ops on these kernels. The size limit prevents constant folding in these cases even though the output of the folding would not have increased the size of the protobuf, and in many cases would actually decrease it by removing the ops.

In TFTRT, TRT requires weights to be known statically so we rely heavily on constant folding. If a weight is not constant folded, we are unable to convert that layer into TRT.

This fixes https://github.com/tensorflow/tensorflow/issues/24083

### CODE REVIEW

**Review by skye — DISMISSED** [review]
Nice, thanks!

**Comment by skye** [comment]
FYI @rmlarsen @ezhulenev

**Comment by trevor-m** [comment]
Sorry, I think I wasn't using the correct input size originally.

There is still a problem where sometimes it thinks the input size is 0 (occuring with Square op in https://github.com/tensorflow/tensorflow/issues/24083)

File: tensorflow/core/grappler/optimizers/constant_folding.cc [hunk 1]
```diff
@@ -1011,6 +1010,7 @@ Status ConstantFolding::EvaluateOneFoldable(const NodeDef& node,
     }
   });
 
+  size_t input_size = 0;
```
Comment (skye) [inline tensorflow/core/grappler/optimizers/constant_folding.cc]: nit: maybe total_inputs_size or similar?

File: tensorflow/core/grappler/optimizers/constant_folding.h [hunk 2]
```diff
@@ -36,7 +36,7 @@ const char kConstantFoldingCtrl[] = "ConstantFoldingCtrl";
 class ConstantFolding : public GraphOptimizer {
  public:
   static Status CreateNodeDef(const string& name, const TensorValue& tensor,
-                              NodeDef* node);
+                              NodeDef* node, size_t input_size = 0);
```
Comment (skye) [inline tensorflow/core/grappler/optimizers/constant_folding.h]: Document what input_size is

**Review by skye — CHANGES_REQUESTED** [review]

**Comment by skye** [comment]
Ah I keep reviewing as you're commenting/making changes :) Are you planning to address the Square op issue in this PR? If so, please let me know when it's ready to review.

**Comment by trevor-m** [comment]
> Ah I keep reviewing as you're commenting/making changes :) Are you planning to address the Square op issue in this PR? If so, please let me know when it's ready to review.

Sorry about that, I discovered the problems shortly after creating the pr. I will let you know after I have investigating the square issue. Thanks for reviewing!

**Comment by trevor-m** [comment]
@skye I have fixed the problem now.

File: tensorflow/core/grappler/optimizers/constant_folding.h [hunk 3]
```diff
@@ -35,8 +35,10 @@ const char kConstantFoldingCtrl[] = "ConstantFoldingCtrl";
 // Constant folding optimization for a graph.
 class ConstantFolding : public GraphOptimizer {
  public:
+  // The size limit will only be considered if the newly created node is greater
+  // than input_size (optional).
```
Comment (skye) [inline tensorflow/core/grappler/optimizers/constant_folding.h]: nit: rename input_size to original_size or something, so it's more clear what this means outside of the context of how you call it.

**Review by skye — DISMISSED** [review]

File: tensorflow/core/grappler/optimizers/constant_folding.cc [hunk 4]
```diff
@@ -980,11 +981,11 @@ Status ConstantFolding::CreateNodeDef(const string& name,
   }
   node->mutable_attr()->insert({"value", attr_tensor});
 
-  if (encoded_size < 10 * 1024 * 1024) {
-    return Status::OK();
+  if (encoded_size > original_size && encoded_size >= 10 * 1024 * 1024) {
+    return errors::InvalidArgument(
+        strings::StrCat("Can't fold ", name, ", its size would be too large"));
```
Comment (pooyadavoodi) [inline tensorflow/core/grappler/optimizers/constant_folding.cc]: I suggest to print both the limit and the size here in case of failure.

**Review by ezhulenev — DISMISSED** [review]

**Comment by ezhulenev** [comment]
Thanks for the fix.

**Review by ezhulenev — APPROVED** [review]
